spec/Spec/Polynomials.cry

```diff
@@ -0,0 +1,53 @@
+/**
+ * The module contains functions that perform BLS polynomial operations.
+ * @see https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/polynomial-commitments.md#polynomials
+ */
+module Spec::Polynomials where
+
+import Common::Utils
+import Spec::BlsHelpers
+import Spec::Permutations
+
+/**
+ * The number of BLS field elements in a data Blob.
+ * @see https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/polynomial-commitments.md#blob
+ */
+type FIELD_ELEMENTS_PER_BLOB = 4096
+
+/**
+ * A polynomial in evaluation form.
+ * @see https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/polynomial-commitments.md#custom-types
+ */
+type BlsPolynomial = [FIELD_ELEMENTS_PER_BLOB]BlsFieldElement
+
+/**
+ * Evaluate a polynomial (in evaluation form) at an arbitrary point `z`.
+ *  - When `z` is in the domain, the evaluation can be found by indexing the polynomial at the
+ *  position that `z` is in the domain.
+ *  - When `z` is not in the domain, the barycentric formula is used:
+ *     f(z) = (z^^WIDTH - 1) / WIDTH  *  sum_(i=0)^WIDTH  (f(DOMAIN[i]) * DOMAIN[i]) / (z - DOMAIN[i])
+ * @see https://github.com/ethereum/consensus-specs/blob/dev/specs/deneb/polynomial-commitments.md#evaluate_polynomial_in_evaluation_form
+ */
+evaluate_polynomial_in_evaluation_form : BlsPolynomial -> BlsFieldElement -> BlsFieldElement
+evaluate_polynomial_in_evaluation_form polynomial z =
+    // If we are asked to evaluate within the domain, we already know the answer
+    if within_domain then
+        polynomial@(roots_of_unity_brp@z)
+    else
+        evaluate_polynomial_not_within_domain
+    where
+        width = `(FIELD_ELEMENTS_PER_BLOB)
+        inverse_width = fromIntegral`{BLS_MODULUS} (bls_modular_inverse width)
+        roots_of_unity_brp = bit_reversal_permutation`{len=FIELD_ELEMENTS_PER_BLOB} (compute_roots_of_unity `(FIELD_ELEMENTS_PER_BLOB))
```

Comment:
I like this comment --- I think I needed to specify it, but I will definitely double check.

Reply:
I cannot get Cryptol to automatically determine this, perhaps because `bit_reversal_permutation` has two type parameters. Perhaps I hard code it to `BlsFieldElements`?

Reply:
Ah, I poked at this. I didn't look carefully the first time, but `compute_roots` has both a type parameter and an argument, which have to be the same value. I parsed it as defining the type parameter to be `FIELD_ELTS_PER_BLOB` but it is actually the argument. So you have to manually put the type (also `FIELD_ELTS_PER_BLOB`) either on `compute_roots` or `bit_reversal`. 

If you wanted to change this I would suggest to drop the argument from `compute_roots` and rename the type parameter to `order`. Then you can put the assertion (order mod modulus) as a type constraint and the rest kind of falls into place. 

I think it's also fine if you don't change this and leave it as-is.

Reply:
The reason I kept it as an argument is because that is how the Python (and C) implementations specify the functions, so I thought it would be easier on the SAW side when it comes to making proofs. I think I'll just leave it this way. Thanks!